### PR TITLE
OCI Auto mode in Singularity/Apptainer

### PR DESCRIPTION
**Pull request by pditommaso**
Add the ability to run OCI containers with Apptainer without the need to convert them explicitly via [ApptainerCache](https://github.com/nextflow-io/nextflow/blob/e9aa8b223ebc1eba9d3627d69f93446fa3004965/modules/nextflow/src/main/groovy/nextflow/container/ApptainerCache.groovy#L28-L28)

### CODE REVIEW

**Comment by marcodelapierre** [comment]
Hold on, see my thread on the originating issue: https://github.com/nextflow-io/nextflow/issues/4543

**Comment by pditommaso** [comment]
To summarise, currently this PR allows the use of OCI (Docker) containers converting implicitly to SIF image file format both for Singularity and Apptainer by adding respectively the option 

```
singularity.direct = true
apptainer.direct = true
```

Along with this is also supported the native OCI support in singularity 4 by using 

```
singularity.oci = true
```

However, considering they are both OCI containers may a better naming could be 

```
apptainer.autoOciMode = true|false
singularity.autoOciMode = true|false
```

and 

```
singularity.nativeOciMode = true|false
```


OR 

```
apptainer.ociAuto = true|false
singularity.ociAuto = true|false
```

and 

```
singularity.ociNative = true|false
```

**Comment by marcodelapierre** [comment]
Yep, second 100% your summary around the proposed functionalities.  
Personally, I like `ociAuto` and `ociNative`, as they are concise.  
Also, not that `ociNative` in Singularity implies `ociAuto`.

I will review this PR on Wed at the latest.

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/resolver/WaveContainerResolver.groovy [hunk 1]
```diff
@@ -70,8 +70,7 @@ class WaveContainerResolver implements ContainerResolver {
         final freeze = client().config().freezeMode()
         final config = task.getContainerConfig()
         final engine = getContainerEngine0(config)
-        final singularityOciMode = config.singularityOciMode()
-        final singularitySpec = freeze && engine in SINGULARITY_LIKE && !singularityOciMode
```
[the remark this answers is not in the document]
Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/resolver/WaveContainerResolver.groovy:74]: OK, I see  that SingularitySpec controls whether a Singularityfile or Dockerfile is used. 
At this stage it is not clear to me why this is conditional to `canRunOciImage` being false; though this is not in scope for this PR.

**Review by marcodelapierre — CHANGES_REQUESTED** [review]
A few small requests along the code - thanks Paolo!

**Comment by marcodelapierre** [comment]
Forgot to mention: these updates to Oci options need to be reflected in the `docs` as well

File: modules/nextflow/src/main/groovy/nextflow/container/ContainerHandler.groovy [hunk 2]
```diff
@@ -67,7 +67,7 @@ class ContainerHandler {
             final normalizedImageName = normalizeSingularityImageName(imageName)
             if( !config.isEnabled() || !normalizedImageName )
                 return normalizedImageName
-            if( normalizedImageName.startsWith('docker://') && config.singularityOciMode() )
+            if( normalizedImageName.startsWith('docker://') && config.canRunOciImage() )
```
[the remark this answers is not in the document]
Comment (pditommaso) [reply modules/nextflow/src/main/groovy/nextflow/container/ContainerHandler.groovy:70]: > I think we can disable image caching in the work dir also in the case of normalized

Don't know. Let's keep for a following iteration

**Comment by pditommaso** [comment]
Ok, considering the term *OCI native* is used is several places with different meaning, I went for `ociMode` (that's how it's [referred in the docs](https://docs.sylabs.io/guides/latest/user-guide/oci_runtime.html#oci-mode-oci)) and `ociAutoPull`. 

The docs has been updated correspondingly.

**Pull request merged**